packages/url_launcher/url_launcher_ios/ios/Classes/URLLauncherPlugin.swift

```diff
@@ -0,0 +1,107 @@
+  private let launcher: Launcher
+
+  var topViewController: UIViewController? {
+    // TODO(stuartmorgan) Provide a non-deprecated codepath. See
+    // https://github.com/flutter/flutter/issues/104117
+    UIApplication.shared.keyWindow?.rootViewController?.topViewController
+  }
+
+  init(launcher: Launcher = UIApplicationLauncher()) {
+    self.launcher = launcher
+  }
+
+  func canLaunchUrl(url: String) throws -> Bool {
+    guard let url = URL(string: url) else {
+      throw invalidURLError(for: url)
+    }
+    return launcher.canOpenURL(url)
+  }
+
+  func launchUrl(
+    url: String, universalLinksOnly: Bool, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+    let options = [UIApplication.OpenExternalURLOptionsKey.universalLinksOnly: universalLinksOnly]
+    launcher.openURL(url, options: options) { success in
+      completion(Result.success(success))
+    }
+  }
+
+  func openUrlInSafariViewController(
+    url: String, completion: @escaping (Result<Bool, Error>) -> Void
+  ) {
+    guard let url = URL(string: url) else {
+      completion(Result.failure(invalidURLError(for: url)))
+      return
+    }
+
+    let session = URLLaunchSession(url: url, completion: completion)
+    currentSession = session
+
+    session.didFinish = { [weak self] in
+      self?.currentSession = nil
+    }
+    topViewController?.present(session.safariViewController, animated: true, completion: nil)
+  }
+
+  func closeSafariViewController() throws {
+    currentSession?.close()
+  }
+
+  /**
+    * Creates an error for an invalid URL string.
+    *
+    * @param url The invalid URL string
+    * @return The error to return
+    */
+  func invalidURLError(for url: String) -> Error {
+    FlutterError(
+      code: "argument_error", message: "Unable to parse URL", details: "Provided URL: \(url)")
+  }
+}
+
+/// This method recursively iterates through the view hierarchy
+/// to return the top-most view controller.
+///
+/// It supports the following scenarios:
+///
+/// - The view controller is presenting another view.
+/// - The view controller is a UINavigationController.
+/// - The view controller is a UITabBarController.
+///
+/// @return The top most view controller.
+extension UIViewController {
+  var topViewController: UIViewController {
+    if let navigationController = self as? UINavigationController {
+      return navigationController.viewControllers.last?.topViewController ?? navigationController
+        .visibleViewController ?? navigationController
+    }
+    if let tabBarController = self as? UITabBarController {
+      return tabBarController.selectedViewController?.topViewController ?? tabBarController
+    }
+    if let presented = presentedViewController {
+      return presented.topViewController
+    }
+    return self
+  }
+}
+
+extension FlutterError: Error {}
```

Comment:
Nice! 

After we make `FlutterError` a subclass of `NSError` in the engine, this will be a compile error complaining about redundant conformance. We will just need to delete this.

Reply:
That's not a feature.

When that happens, how will `url_launcher` support both `master` and `stable`, as we require for our packages?

Packages don't work like the engine or the framework, in that they have to be compatible with more than exactly one version of all of the engine code, and they have to be designed differently as a result. If collisions are a compile error, then we shouldn't do this.

Reply:
Relatedly, https://github.com/flutter/packages/pull/4753#discussion_r1349219420 still hasn't been addressed.